### PR TITLE
Refactor base profilers 3/5

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Introduce Abstract interface
- Clean up logic and inheritance
- Change `output_fname` for `dirpath` + `filepath`
(proposed in https://github.com/PyTorchLightning/pytorch-lightning/pull/6349/files/6f2258f5d07c919cd9b7b040d8cb52e630cd4c62#r598366936)
- Allow report files per stage and rank with the `{stage}-filename-{rank}.txt` format

Fixes https://github.com/PyTorchLightning/pytorch-lightning/pull/6522
Note that this fix will not be included in a bug-fix release as it requires the whole profiler refactor

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/profiler/profilers.py [hunk 1]
```diff
@@ -55,6 +41,49 @@ def start(self, action_name: str) -> None:
...
+        and `TensorBoardLogger` log_dir in the profiler.
+        """
+        self.local_rank = local_rank
+        self.log_dir = log_dir
+
+    def _prepare_file(self) -> None:
+        if not self._file_prepared:
+            if self.output_fname and self.output_file is None:
+                fs = get_filesystem(self.output_fname)
+                self.output_file = fs.open(self.output_fname, "w")
```
Comment (ananthsub) [inline pytorch_lightning/profiler/profilers.py:83]: are we only appending to files?
```suggestion
                self.output_file = fs.open(self.output_fname, "a+")
```

File: pytorch_lightning/profiler/profilers.py [hunk 2]
```diff
@@ -88,15 +117,41 @@ def profile_iterable(self, iterable, action_name: str) -> None:
 
     def describe(self) -> None:
         """Logs a profile report after the conclusion of the training run."""
+        self._prepare_file()
         for write in self.write_streams:
             write(self.summary())
+        if self.output_file:
+            self.output_file.flush()
+        self.teardown()
```
Comment (ananthsub) [inline pytorch_lightning/profiler/profilers.py:125]: `teardown` could be called directly by the trainer during the teardown hooks. do we want to nest it inside of `describe` ?

File: pytorch_lightning/profiler/profilers.py [hunk 3]
```diff
@@ -88,15 +117,41 @@ def profile_iterable(self, iterable, action_name: str) -> None:
...
-    def on_train_start(self, local_rank: Optional[int] = None):
-        self.local_rank = local_rank
+    def start(self, action_name: str) -> None:
+        raise NotImplementedError
+
+    def stop(self, action_name: str) -> None:
+        raise NotImplementedError
+
+    def summary(self) -> str:
+        raise NotImplementedError
```
Comment (ananthsub) [inline pytorch_lightning/profiler/profilers.py:154]: n00b q: why are these needed? would someone directly instantiate the `BaseProfiler`?

Comment (carmocca) [reply pytorch_lightning/profiler/profilers.py:154]: To avoid the IDE warning `Class BaseProfiler must implement all abstract methods `

File: pytorch_lightning/profiler/profilers.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/profiler/profilers.py]: I think `a` is okay, but why `a+`? Do we expect anybody to rely on this handle to read?

File: pytorch_lightning/profiler/profilers.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/profiler/profilers.py:114]: @tchaton I believe you added it. Any reason to have it here? We already have it in:

https://github.com/PyTorchLightning/pytorch-lightning/blob/01a760e7a42ae0686a67ca80b5f06188e5f7f9f7/pytorch_lightning/trainer/trainer.py#L1077-L1085

File: tests/test_profiler.py [hunk 4]
```diff
@@ -316,3 +316,21 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
...
+
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        fast_dev_run=True,
+        profiler=profiler,
+    )
+    trainer.fit(model)
+
+    assert profiler.output_file.closed
```
Comment (ananthsub) [inline tests/test_profiler.py:336]: one thing i'm concerned about is if someone does this:
```
trainer = Trainer(..., profiler=Profiler(...))
trainer.fit(module)
trainer.test(module)
```
with the setup and teardown hooks, will the `test` call overwrite the results from the `fit` call?

do we want to approach this with the proposal from #6420 and introduce additional hooks into the profiler API?

if you think that would bloat the profiler API, we could add some handling to include the stage information in the filename to avoid collisions, and then add versioning on top of that so we don't silently overwrite things

File: pytorch_lightning/profiler/profilers.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/profiler/profilers.py]: yes you're right, it can just be `a`

File: tests/test_profiler.py (hunk 4 above)
Comment (carmocca) [reply tests/test_profiler.py:336]: Good point!

> do we want to approach this with the proposal from #6420 and introduce additional hooks into the profiler API?

I think I'd rather let other people chime in on that proposal instead of implementing it already for this. We also need to move fast so that would be more overhead for reviewers

So for now, we can take the `stage` argument in both `setup` and `teardown`, and do as you propose:

> include the stage information in the filename to avoid collisions, and then add versioning on top of that so we don't silently overwrite things

File: pytorch_lightning/profiler/profilers.py [hunk 5]
```diff
@@ -117,37 +189,41 @@ def stop(self, action_name: str) -> None:
...
         self.extended = extended
-
-        self.output_fname = output_filename
-        self.output_file = None
-        if self.output_fname:
-            fs = get_filesystem(self.output_fname)
-            self.output_file = fs.open(self.output_fname, "w")
-
-        streaming_out = [self.output_file.write] if self.output_file else [log.info]
+        super().__init__(dirpath=dirpath, filename=filename, output_filename=output_filename)
```
Comment (ananthsub) [inline pytorch_lightning/profiler/profilers.py:225]: nit: move to first line in constructor?

File: pytorch_lightning/profiler/profilers.py [hunk 6]
```diff
@@ -117,37 +189,41 @@ def stop(self, action_name: str) -> None:
...
+        filename: Optional[str] = None,
+        extended: bool = True,
+        output_filename: Optional[str] = None,
+    ) -> None:
         """
         Args:
-            output_filename: optionally save profile results to file instead of printing
-                to std out when training is finished.
+            dirpath: Directory path for the ``filename``. If ``dirpath`` is ``None`` but ``filename`` is present, the
+                ``trainer.log_dir`` (from :class:`~pytorch_lightning.loggers.tensorboard.TensorBoardLogger`)
```
Comment (ananthsub) [inline pytorch_lightning/profiler/profilers.py:212]: nit: this isn't specific to tensorboard right?

Comment (carmocca) [reply pytorch_lightning/profiler/profilers.py:212]: I believe only TensorBoardLogger and CSVLogger define log_dir

File: tests/test_profiler.py (hunk 4 above)
Comment (ananthsub) [reply tests/test_profiler.py]: sgtm - i agree #6420 shouldn't block this and if that's agreed on, we'll need to audit and fix systematically anyways.

File: pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -1068,6 +1068,9 @@ def call_setup_hook(self, model: LightningModule) -> None:
             if not called:
                 self.datamodule.setup(stage=state)
 
+        local_rank = self.local_rank if self.world_size > 1 else None
+        self.profiler.setup(local_rank=local_rank, log_dir=self.log_dir)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1072]: the test failures this needs to be called after L1074 since accessing `log_dir` triggers a broadcast and the process groups aren't set up yet? https://github.com/PyTorchLightning/pytorch-lightning/blob/870247ffe6a5ce819cf7e0a22b997a765d2f6675/pytorch_lightning/trainer/properties.py#L130-L138

but the accelerator environment [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/870247ffe6a5ce819cf7e0a22b997a765d2f6675/pytorch_lightning/trainer/trainer.py#L433
) should call 
- https://github.com/PyTorchLightning/pytorch-lightning/blob/870247ffe6a5ce819cf7e0a22b997a765d2f6675/pytorch_lightning/accelerators/accelerator.py#L72-L78
- https://github.com/PyTorchLightning/pytorch-lightning/blob/870247ffe6a5ce819cf7e0a22b997a765d2f6675/pytorch_lightning/plugins/training_type/ddp.py#L83-L91
- https://github.com/PyTorchLightning/pytorch-lightning/blob/870247ffe6a5ce819cf7e0a22b997a765d2f6675/pytorch_lightning/plugins/training_type/ddp.py#L164-L190
- https://github.com/PyTorchLightning/pytorch-lightning/blob/870247ffe6a5ce819cf7e0a22b997a765d2f6675/pytorch_lightning/plugins/training_type/ddp.py#L234-L242 

is this a race? also triggering a broadcast for each property access can be unexpectedly expensive.

separately, should we should move the datamodule setup to be right before the model setup?

File: pytorch_lightning/profiler/profilers.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/profiler/profilers.py:137]: Hey,

If a user was to perform a deepcopy in a hook between the describe call and the teardown call, 
it would fail in python 3.6 as write_streams can't be pickled.

IMO, I think it is safer to close the write_streams as soon as being used.

File: pytorch_lightning/trainer/trainer.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1072]: Yes, this would be an issue,

A broadcast is needed as only Dummy Logger are created on rank != 0 and they need to be aware of log_dir from rank 0.

```
broadcast for each property access can be unexpectedly expensive
```

@carmocca I think we should introduce a `on_run_stage` hook and change training_type_plugins to call run_stage directly. This will ensure the local_rank is available and we could call `self.profiler.setup` inside.
```
    def on_run_stage(self):
        self.call_hook("on_run_stage")
        self.call_hook(f"on_run_{self._running_stage.value}")

    def run_stage(self):
        results = None
        self.on_run_stage()
        if self.evaluating:
            results = self.run_evaluate()
        elif self.predicting:
            results = self.run_predict()
        else:
            self.run_train()
        return results
```

Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:1072]: could we set the `log_dir` on the dummy logger whenever they're created without needing the `on_run_stage` ?

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1080]: We still need to `on_run_stage_setup` as local_rank needs to be defined when running with ddp_spawn.
Actually, this hooks is needed to ensure everything is properly setup by all training type plugins.

For `log_dir`, It seems Dummy Logger is used only when `fast_dev_run`. Asking @awaelchli about the broadcast.

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -516,8 +516,14 @@ def dispatch(self):
         else:
             self.accelerator.start_training(self)
 
+    def on_run_stage_setup(self):
+        self.profiler.setup(local_rank=self.local_rank if self.world_size > 1 else None, log_dir=self.log_dir)
+
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:521]: if it's not meant to be called outside the trainer
```suggestion
    def _on_run_stage_setup(self):
        self.profiler.setup(local_rank=self.local_rank if self.world_size > 1 else None, log_dir=self.log_dir)

```

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:521]: Good catch ! Definitely.

File: pytorch_lightning/profiler/profilers.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/profiler/profilers.py:139]: After discussing, we chose to keep it as is. Will add a comment mentioning the pickle limitation.

File: pytorch_lightning/profiler/profilers.py [hunk 9]
```diff
@@ -55,23 +43,47 @@ def start(self, action_name: str) -> None:
...
         """Defines how to record the duration once an action is complete."""
 
-    def setup(
+    @abstractmethod
+    def summary(self) -> str:
+        """Create profiler summary in text format."""
+
+    @abstractmethod
+    def setup(self) -> None:
+        """Execute arbitrary pre-profiling set-up steps as defined by subclass."""
```
Comment (ananthsub) [inline pytorch_lightning/profiler/profilers.py:52]: should setup take **kwargs?

File: pytorch_lightning/profiler/profilers.py [hunk 10]
```diff
@@ -55,23 +43,47 @@ def start(self, action_name: str) -> None:
...
+
+    @abstractmethod
+    def teardown(self) -> None:
+        """Execute arbitrary post-profiling tear-down steps as defined by subclass."""
```
Comment (ananthsub) [inline pytorch_lightning/profiler/profilers.py:56]: same for teardown?

File: pytorch_lightning/profiler/profilers.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/profiler/profilers.py:52]: Do you see any reasons for it ?

Comment (ananthsub) [reply pytorch_lightning/profiler/profilers.py:52]: The base class overrides this differently by adding more args

Comment (tchaton) [reply pytorch_lightning/profiler/profilers.py:52]: @carmocca Let's add kwargs there and below.

File: pytorch_lightning/profiler/profilers.py [hunk 11]
```diff
@@ -145,30 +220,32 @@ class SimpleProfiler(BaseProfiler):
     the mean duration of each action and the total time spent over the entire training run.
     """
 
-    def __init__(self, output_filename: Optional[str] = None, extended=True):
+    def __init__(
+        self,
+        dirpath: Optional[Union[str, Path]] = None,
```
Comment (tchaton) [inline pytorch_lightning/profiler/profilers.py:225]: dirpath and filename is not backward compatible. Any idea to better handle this ?

File: pytorch_lightning/trainer/evaluation_loop.py [hunk 12]
```diff
@@ -99,6 +100,10 @@ def on_evaluation_end(self, *args, **kwargs):
         else:
             self.trainer.call_hook('on_validation_end', *args, **kwargs)
 
+        if self.trainer.state != TrainerState.FITTING:
+            # summarize profile results
+            self.trainer.profiler.describe()
```
Comment (tchaton) [inline pytorch_lightning/trainer/evaluation_loop.py:105]: Let's add a there for test and validate. Don't we call describe twice for test ?

**Review by tchaton — APPROVED** [review]
LGTM ! small comments.

File: pytorch_lightning/profiler/profilers.py (hunk 11 above)
Comment (carmocca) [reply pytorch_lightning/profiler/profilers.py:225]: we still have `output_filename` for BC

File: pytorch_lightning/trainer/evaluation_loop.py (hunk 12 above)
Comment (carmocca) [reply pytorch_lightning/trainer/evaluation_loop.py:105]: done

**Review by ananthsub — APPROVED** [review]

File: CHANGELOG.md [hunk 13]
```diff
@@ -51,6 +51,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added `Trainer.predict` config validation ([#6543](https://github.com/PyTorchLightning/pytorch-lightning/pull/6543))
 
 
+- Added `AbstractProfiler` interface ([#6621](https://github.com/PyTorchLightning/pytorch-lightning/pull/6621))
+
+
+- Added `setup` method to `BaseProfiler` to enable subclasses defining pre-profiling steps ([#6621](https://github.com/PyTorchLightning/pytorch-lightning/pull/6621))
+
+
```
Comment (carmocca) [inline CHANGELOG.md:59]:
```suggestion
```

File: pytorch_lightning/profiler/profilers.py [hunk 14]
```diff
@@ -55,23 +43,47 @@ def start(self, action_name: str) -> None:
...
+    @abstractmethod
+    def setup(self, **kwargs: Any) -> None:
+        """Execute arbitrary pre-profiling set-up steps as defined by subclass."""
+
+    @abstractmethod
+    def teardown(self, **kwargs: Any) -> None:
+        """Execute arbitrary post-profiling tear-down steps as defined by subclass."""
+
+
+class BaseProfiler(AbstractProfiler):
```
Comment (kaushikb11) [inline pytorch_lightning/profiler/profilers.py:59]: @tchaton Why do we need `BaseProfiler` & `AbstractProfiler` both?

File: tests/trainer/properties/test_get_model.py [hunk 15]
```diff
@@ -84,8 +83,7 @@ def test_get_model_gpu(tmpdir):
 
 
 @RunIf(min_gpus=1, skip_windows=True)
-@DDPLauncher.run("--accelerator [accelerator]", max_epochs=["1"], accelerator=["ddp", "ddp_spawn"])
-def test_get_model_ddp_gpu(tmpdir, args=None):
+def test_get_model_ddp_gpu(tmpdir):
```
Comment (justusschock) [inline tests/trainer/properties/test_get_model.py:86]: Why don't we need the DDPLauncher here any more?

**Review by justusschock — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

File: tests/trainer/properties/test_get_model.py (hunk 15 above)
Comment (tchaton) [reply tests/trainer/properties/test_get_model.py:86]: We never needed it :)

File: pytorch_lightning/profiler/profilers.py (hunk 14 above)
Comment (tchaton) [reply pytorch_lightning/profiler/profilers.py:59]: I think slightly cleaner.

File: tests/trainer/properties/test_get_model.py (hunk 15 above)
Comment (awaelchli) [reply tests/trainer/properties/test_get_model.py:86]: it doesn't even do anything.
this test can be deleted, now that not even the name has a meaning anymore

File: pytorch_lightning/profiler/profilers.py [hunk 16]
```diff
@@ -104,19 +116,94 @@ def profile_iterable(self, iterable, action_name: str) -> None:
...
-    @abstractmethod
     def summary(self) -> str:
-        """Create profiler summary in text format."""
+        raise NotImplementedError
 
-    def __del__(self):
-        self.teardown(None)
+    @property
+    def local_rank(self):
+        return '0' if self._local_rank is None else self._local_rank
```
Comment (awaelchli) [inline pytorch_lightning/profiler/profilers.py:206]: is there a special meaning for the string? _local_rank is annotated as int
one should be especially careful with comparison here
